### PR TITLE
Make Rounding.nextRoundingValue consistent

### PR DESCRIPTION
**Pull request by nik9000**
"interval" style roundings were implementing `nextRoundingValue` in a
fairly inconsistent way - it'd produce a value, but sometimes that
value would be the same as the previous rounding value. This makes it
consistently the next value that `rounding` would make.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/common/RoundingTests.java [hunk 1]
```diff
@@ -469,6 +469,20 @@ public void testRandomTimeIntervalRounding() {
...
+        long next = rounding.prepareForUnknown().nextRoundingValue(rounded);
+        assertThat(next, greaterThan(rounded));
+    }
+
+    public void testBar() {
+        Rounding rounding = new Rounding.TimeIntervalRounding(480000, ZoneId.of("Portugal"));
+        long rounded = rounding.prepareJavaTime().round(972780720000L);
+        long next = rounding.prepareJavaTime().nextRoundingValue(rounded);
+        assertThat(next, greaterThan(rounded));
+    }
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/RoundingTests.java:484]: I'm not sure I need to keep these. They failed on some of my other attempts at implementing this method.

Comment (not-napoleon) [reply server/src/test/java/org/elasticsearch/common/RoundingTests.java:484]: I'd keep them in, personally.  Could just put a note that those specific values caused problems for some implementations, so we're keeping them around for regression testing, or something like that.

File: server/src/main/java/org/elasticsearch/common/Rounding.java [hunk 2]
```diff
@@ -1119,12 +1118,16 @@ public long round(long utcMillis) {
...
 
             @Override
-            public long nextRoundingValue(long time) {
-                int offsetSeconds = timeZone.getRules().getOffset(Instant.ofEpochMilli(time)).getTotalSeconds();
-                long millis = time + interval + offsetSeconds * 1000;
-                return ZonedDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneOffset.UTC)
-                    .withZoneSameLocal(timeZone)
-                    .toInstant().toEpochMilli();
+            public long nextRoundingValue(long utcMillis) {
+                long from = utcMillis + interval; // TODO this implementation makes me very upset
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/common/Rounding.java:1122]: So I'm just glancing at this, and haven't read the whole class, but, it looks like the old implementation was using `interval` as a number of seconds, and this is using it as a number of milliseconds?  Is that correct?  Or am I misunderstanding something here?

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/common/Rounding.java:1122]: The old implementation had the zone's offset in seconds but the interval was in millis.

**Comment by nik9000** [comment]
I ran the rounding benchmarks on this and it makes `nextRoundingValue` about 10x slower:
```
before
(count)     (interval)                   (range)  (rounder)            (zone)  Mode  Cnt          Score          Error  Units
1000000            10d  2000-01-01 to 2020-01-01  java time               UTC  avgt   10   77222270.084 ±  1629956.706  ns/op
1000000            10d  2000-01-01 to 2020-01-01  java time  America/New_York  avgt   10  148164347.557 ±   366215.848  ns/op
1000000            10d  2000-10-01 to 2000-11-01  java time               UTC  avgt   10   69270374.524 ± 13644732.543  ns/op
1000000            10d  2000-10-01 to 2000-11-01  java time  America/New_York  avgt   10  115496849.720 ± 10201433.591  ns/op
1000000            10d  2000-10-29 to 2000-10-30  java time               UTC  avgt   10   77867390.969 ±   136230.473  ns/op
1000000            10d  2000-10-29 to 2000-10-30  java time  America/New_York  avgt   10  125616344.303 ±  4890854.395  ns/op
1000000            10d  2000-06-01 to 2000-06-02  java time               UTC  avgt   10   78933290.977 ±  1854368.454  ns/op
1000000            10d  2000-06-01 to 2000-06-02  java time  America/New_York  avgt   10  100506143.585 ±  8384023.616  ns/op
1000000             5d  2000-01-01 to 2020-01-01  java time               UTC  avgt   10   76424295.586 ±   205871.825  ns/op
1000000             5d  2000-01-01 to 2020-01-01  java time  America/New_York  avgt   10  145355467.816 ± 11192085.570  ns/op
1000000             5d  2000-10-01 to 2000-11-01  java time               UTC  avgt   10   78916422.115 ±  1545254.839  ns/op
1000000             5d  2000-10-01 to 2000-11-01  java time  America/New_York  avgt   10  101937477.980 ±   298725.209  ns/op
1000000             5d  2000-10-29 to 2000-10-30  java time               UTC  avgt   10   77930453.546 ±   146456.658  ns/op
1000000             5d  2000-10-29 to 2000-10-30  java time  America/New_York  avgt   10  120373334.456 ±  1727215.958  ns/op
1000000             5d  2000-06-01 to 2000-06-02  java time               UTC  avgt   10   69205872.803 ± 13866278.871  ns/op
1000000             5d  2000-06-01 to 2000-06-02  java time  America/New_York  avgt   10   95767997.327 ±  1153456.736  ns/op
1000000             1h  2000-01-01 to 2020-01-01  java time               UTC  avgt   10   72298353.129 ±   221990.422  ns/op
1000000             1h  2000-01-01 to 2020-01-01  java time  America/New_York  avgt   10  147658089.243 ±  1366048.284  ns/op
1000000             1h  2000-10-01 to 2000-11-01  java time               UTC  avgt   10   77677151.254 ±   105619.931  ns/op
1000000             1h  2000-10-01 to 2000-11-01  java time  America/New_York  avgt   10  122034758.658 ±  4388995.373  ns/op
1000000             1h  2000-10-29 to 2000-10-30  java time               UTC  avgt   10   79132470.792 ±  1126975.166  ns/op
1000000             1h  2000-10-29 to 2000-10-30  java time  America/New_York  avgt   10  129100779.547 ±  7603475.857  ns/op
1000000             1h  2000-06-01 to 2000-06-02  java time               UTC  avgt   10   78239039.138 ±   371485.769  ns/op
1000000             1h  2000-06-01 to 2000-06-02  java time  America/New_York  avgt   10  114913994.677 ± 12783699.403  ns/op


after
(count)     (interval)                   (range)  (rounder)            (zone)  Mode  Cnt           Score          Error  Units
1000000            10d  2000-01-01 to 2020-01-01  java time               UTC  avgt   10   395324193.500 ± 18645421.073  ns/op
1000000            10d  2000-01-01 to 2020-01-01  java time  America/New_York  avgt   10   833026071.950 ±  8244228.451  ns/op
1000000            10d  2000-10-01 to 2000-11-01  java time               UTC  avgt   10   373163631.667 ± 25070561.888  ns/op
1000000            10d  2000-10-01 to 2000-11-01  java time  America/New_York  avgt   10   833413984.250 ±  9936919.205  ns/op
1000000            10d  2000-10-29 to 2000-10-30  java time               UTC  avgt   10   349217668.150 ± 28665144.075  ns/op
1000000            10d  2000-10-29 to 2000-10-30  java time  America/New_York  avgt   10  1349883838.200 ±  4370203.151  ns/op
1000000            10d  2000-06-01 to 2000-06-02  java time               UTC  avgt   10   365661391.100 ± 35817907.626  ns/op
1000000            10d  2000-06-01 to 2000-06-02  java time  America/New_York  avgt   10   547628071.800 ±  1016719.087  ns/op
1000000             5d  2000-01-01 to 2020-01-01  java time               UTC  avgt   10   373133903.133 ± 13170105.580  ns/op
1000000             5d  2000-01-01 to 2020-01-01  java time  America/New_York  avgt   10   805718088.550 ±  4599848.541  ns/op
1000000             5d  2000-10-01 to 2000-11-01  java time               UTC  avgt   10   325229160.950 ±  8105838.843  ns/op
1000000             5d  2000-10-01 to 2000-11-01  java time  America/New_York  avgt   10   739509291.500 ±  1647471.583  ns/op
1000000             5d  2000-10-29 to 2000-10-30  java time               UTC  avgt   10   359565851.858 ± 45669897.947  ns/op
1000000             5d  2000-10-29 to 2000-10-30  java time  America/New_York  avgt   10  1324330322.200 ±  5610229.568  ns/op
1000000             5d  2000-06-01 to 2000-06-02  java time               UTC  avgt   10   360042488.567 ± 45576771.998  ns/op
1000000             5d  2000-06-01 to 2000-06-02  java time  America/New_York  avgt   10   566384899.300 ± 33740853.436  ns/op
1000000             1h  2000-01-01 to 2020-01-01  java time               UTC  avgt   10   411730394.400 ±  2974945.222  ns/op
1000000             1h  2000-01-01 to 2020-01-01  java time  America/New_York  avgt   10   835122322.800 ±   563651.507  ns/op
1000000             1h  2000-10-01 to 2000-11-01  java time               UTC  avgt   10   330776239.200 ±   274415.757  ns/op
1000000             1h  2000-10-01 to 2000-11-01  java time  America/New_York  avgt   10   623367626.200 ±  1193587.192  ns/op
1000000             1h  2000-10-29 to 2000-10-30  java time               UTC  avgt   10   331757691.925 ±  1568299.645  ns/op
1000000             1h  2000-10-29 to 2000-10-30  java time  America/New_York  avgt   10   654410588.400 ±  3969091.038  ns/op
1000000             1h  2000-06-01 to 2000-06-02  java time               UTC  avgt   10   360812824.783 ± 47793621.994  ns/op
1000000             1h  2000-06-01 to 2000-06-02  java time  America/New_York  avgt   10   583533375.700 ± 57733211.175  ns/op
```

This is likely to be fine thought because we very rarely call this in a hot loop. The only real time we'll see it is when you do a `date_histogram` on a range field. And we can likely speed it up significantly using the optimizations that this unlocks.

File: server/src/main/java/org/elasticsearch/common/Rounding.java [hunk 3]
```diff
@@ -1119,12 +1122,60 @@ public long round(long utcMillis) {
...
+                        continue;
+                    }
+                    assert highEnough && false == tooHigh;
+                    assert roundedRoundedDown == prevRound;
+                    if (iterations > 3 && logger.isDebugEnabled()) {
+                        logger.debug("Iterated {} time for {} using {}", iterations, utcMillis, TimeIntervalRounding.this.toString());
+                    }
+                    return rounded;
+                }
+                assert false : String.format(
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/common/Rounding.java:1167]: I think you're using `assert false` here so this throws in testing but not in production.  If that's the case, please just put a note for why we want the behavior to be different in prod, and maybe what will actually be returned in that case.  And if not, please throw an actual exception.

**Review by not-napoleon — APPROVED** [review]
LGTM.  Nice work on a tricky problem!

File: server/src/main/java/org/elasticsearch/common/Rounding.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/common/Rounding.java:1167]: Sure!

**Pull request merged**